### PR TITLE
[5.0] Rectify the logic of calculating item range

### PR DESCRIPTION
**Pull request by CarterZhou**
[5.0 ]Rectify the logic of calculating item range of a specific page.

### CODE REVIEW

**Comment by taylorotwell** [comment]
Needs explanation of what is wrong with current version and how your fix fixes it.

**Comment by Arrilot** [comment]
The issue is described here https://github.com/laravel/framework/issues/7986

File: src/Illuminate/Pagination/AbstractPaginator.php [hunk 1]
```diff
@@ -232,7 +232,7 @@ public function items()
 	 */
 	public function firstItem()
 	{
-		return ($this->currentPage - 1) * $this->perPage + 1;
+		return $this->count() == 0 ? $this->count() : ($this->currentPage - 1) * $this->perPage + 1;
```
Comment (Arrilot) [inline src/Illuminate/Pagination/AbstractPaginator.php]: I would simplify this to
```php
return $this->count() ? ($this->currentPage - 1) * $this->perPage + 1 : 0;
```
but it doesn't really matter

**Comment by phanan** [comment]
@CarterZhou Correct me if I'm wrong, but with this "fix," will your text in #7986 become **Showing item 0 to 0**, which is meaningless anyway? Shouldn't this check be done in your app logic instead?

**Comment by CarterZhou** [comment]
@phanan Being meanless or not depends on how you perceive the problem. It's minor, but still an issue.

As I said in #7986 , if no data is retrieved,using current version of firstItem() return 1, and lastItem() returns 0. The message could be "showing item 1 to 0", which is also meaningless, right?

So if there is no item at all,  then the first item should not item 1.

Yes, you can do some checking in the view like:
```php
@if($users->count() > 0)
<!-- show item range -->
@else
<!-- show nothing -->
@endif
```
But from my perspective, this's just hiding the incorrect item range, right?

**Comment by CarterZhou** [comment]
@taylorotwell 
I don't think firstItem() should indicate that the first item is item 1, if there is no data in the slice at all.

Therefore in my fix I added checking for empty data set: if there is no item in the slice, return 0 instead; otherwise, using the existing logic.

And because firstItem() is changed, lastItem(), which currently calls firstItem() in its body, needs to change as well. 
No calling firstItem() anymore, but calculates the last item number based on current page number and the number of items in current slice.

**Comment by phanan** [comment]
Actually I think it's wrong to say "the first position of an empty range is
0" (what you're doing) too. If you knowingly pass incorrect data, how can
you expect correct values to be returned? I'd go for an
InvalidArgumentException
instead.

**Pull request closed without merging**